docs/RFCS/00000000_cascade.md

```diff
@@ -0,0 +1,1210 @@
+- `UC` should be fetched and not walked any further.
+- `US` should be fetched and not walked any further.
+
+###### Update
+
+When updating a row in `X`, similar to the insert case, all the tables that table `X` references are required. Also similar to the insert case, there is also no need for those to be walked any further. So `'NA`, `'DC`, `'DS`, `'UC` and `'US` will be fetched and not walked.
+
+For the tables that reference `X`, an update requires the following:
+
+- `NA` should be fetched and not walked any further.
+- `DC` should be fetched and not walked any further.
+- `DU` should be fetched and not walked any further.
+- `UC` should be fetched and walked as an `UPDATE` operation.
+- `US` should be fetched and walked as an `UPDATE` operation.
+
+##### `TablesNeededForFKs`
+
+`fk.go`'s `TablesNeededForFKs` function will have to be expanded. Currently, it only fetches the table IDs that need to be checked when an update, delete or insert operation occurs. With the addition of cascading referential constraints, this is no longer sufficient. Instead of just fetching IDs, this function will now also support optionally retrieving the table descriptors themselves. Without the actual table descriptors, it would not be possible to walk the dependency graph and thus impossible to know if there may be yet another cascading action. This will be done iteratively and should be relatively quick. Care will be taken to not slow down non-cascading examples as this codepath is used with all basic operations. It's important to note that in most operations (i.e. non-backfill ones), immediately after calling `TablesNeededForFKs` lookup all the required table descriptors. The real change here is that more table descriptors will have to be fetched for each operation.
+
+The only augmentation to the function call itself will be to add a passed in function that will perform the actual table descriptor lookup.
+
+The table walk will be accomplished iteratively using a queue and perform the dependency walking algorithm described above. As the final output is a map of table IDs to table descriptors we already have a data structure that can be used for cycle detection. However, we also have to be conscious of the fact that there may be a need to iterate over the same table twice, once with an `update` and once later with a `delete` (or vice versa) due to potential dependency cycles within the graph. This adds a slight complication to the map but can be accounted for.
+
+Both memory and runtime of this function will be bounded based on the number of tables that may potentially be part of a series of cascade operations.
+
+##### `DeleteRow` and `UpdateRow`
+
+Both of `rowwriter.go`'s `DeleteRow` and `UpdateRow` will require a small split so that there are different call sites for when a cascade is desired or not. This will have no memory impact but will obviously, have a performance impact due to the cascading call itself.
+
+##### `Cascader`'s `CascadeAll`
+
```

Comment:
We can call this "execution". I'd propose splitting up the implementation into "planning" and "execution"